File: assignment5/monte1.R
func_rand <- function(val, m){
	return(val/m);
}

func_val <- function(val, a, b, m){
	next_val <- ((a * val) + b) %% m;
	return(next_val);
}

func <- function(initial, a, b, m, end, lambda){
	arr <- array(end);
	cou <- array(end);

	val <- initial;
	random <- func_rand(val, m);

	for(count in 1:end){
		val <- func_val(val, a, b, m);
		random <- func_rand(val, m);
		arr[count] <- random;
	}
	
	sum <- 0.0;
	min <- arr[1];
	max <- arr[1];
	
	for(index in 1:end){
		arr[index] <- -log(1 - arr[index])/lambda;
		cat(sprintf("%g\n", arr[index]));
		
		sum <- sum + arr[index];
		
		if(arr[index] < min){
			min <- arr[index];
		}
			
		if(arr[index] > max){
			max <- arr[index];
		}
	}
	
	mean <- sum/end;

	cat(sprintf("Min=%g\nMax=%g\nMean=%g\n", min, max, mean));
	
	hist(arr, 100,freq = TRUE, include.lowest = TRUE, right = TRUE, density = NULL,
		angle = 45, col = '#1E90FF', border = NULL,
		main = paste("Histogram of frequencies and total values
		generated=",end), xlim=range(0,max), ylim=range(0,(500)), xlab = 'Random Values' ,
		axes = TRUE, plot = TRUE, labels = FALSE);

	dev.copy(jpeg,"hist.jpg");
	dev.off();	
}

main<-function(){
	
	init_value <- 3452;
	bound <- 5000;
	lambda <- 0.2;

	func(init_value, 4696, 0, 5003, bound, lambda);
}

File: assignment4/monte1.cpp
#include <iostream>
#include <cstdio>
#include <cmath>

#define LL long long

using namespace std;

double func_rand(LL int val, LL int m){
	return (double)val/m;
}

LL int func_val(LL int val, LL int a, LL int b, LL int m){
	LL next_val;
	next_val = ((a * val) + b) % m;

	return next_val;
}

void function(LL int initial, LL int a, LL int b, LL int m, int end, double array[]){
	LL int val, count = 0;
	LL int index, quot;
	double random, prev;

	val = initial;
	random = func_rand(val, m);

	while(count<end){
		val = func_val(val, a, b, m);
		random = func_rand(val, m);
		
		array[count] = random;
		++count;
	}
	
	for(index = 0 ; index<20 ; ++index){
		printf("\"%.2lf\", %lld\n", 0.05*(index+1), array[index]);
	}
}

void exponential(double lambda, double array[], LL int bound){
	int index;
	double quot, min, max, mean, sum = 0.0;
	LL int count[100] = {0};
	
	min = max = array[0];
	for(index=0; index<bound; ++index){
		array[index] = -(double)log(1.0 - array[index])/lambda;
		printf("%lf\n", array[index]);
		
		++count[(int)array[index]];
		sum+=array[index];
		
		if(array[index] < min)
			min = array[index];
			
		if(array[index] > max)
			max = array[index];
	}
	printf("\n");
	
	mean = (double)sum/bound;

	printf("Min=%lf\nMax=%lf\nMean=%lf\n", min, max, mean);
	
	for(index = 0; index<=(int)max; ++index){
		printf("\"%d\" , %lld\n", index, count[index]);
	}
}

int main(){
	LL int index;
	double array[5005];
	LL int init_value = 3452;
	LL int bound = 5000;
	double lambda = 0.2;

	function(init_value, 4696, 0, 5003, bound, array);
	exponential(lambda, array, bound);
	printf("\n");

	return 0;
}
